File: frontend/src/services/agentRemoteTarget.ts
import { apiClient } from "../utils/apiClient";

const API = "/api/v1/agents";

export interface RemoteTarget {
    id: number;
    agent_id: number;
    name: string;
    hostname: string;
    protocol: "psremoting" | "ssh" | "winrm";
    port: number;
    username: string;
    enabled: boolean;
    tags: string | null;
    target_plugins: string | null;
    notes: string | null;
    last_collected_at: string | null;
    last_status: string;
    last_error: string | null;
    created_at: string;
    updated_at: string;
}

export interface RemoteTargetCreate {
    name: string;
    hostname: string;
    protocol: string;
    port?: number;
    username: string;
    password?: string;
    ssh_key?: string;
    enabled?: boolean;
    tags?: string;
    target_plugins?: string;
    notes?: string;
}

export interface RemoteTargetUpdate {
    name?: string;
    hostname?: string;
    protocol?: string;
    port?: number;
    username?: string;
    password?: string;
    ssh_key?: string;
    enabled?: boolean;
    tags?: string;
    target_plugins?: string;
    notes?: string;
}

export interface RemoteInventory {
    agent_id: number;
    agent_name: string;
    hostname: string;
    remote_targets: Record<string, {
        hostname: string;
        protocol: string;
        collected_at: string | null;
        status: string;
        error: string | null;
        inventory: Record<string, unknown>;
    }>;
}

export const agentRemoteTargetApi = {
    async listTargets(agentId: number): Promise<RemoteTarget[]> {
        const data = await apiClient<{ targets: RemoteTarget[] }>(
            `${API}/${agentId}/remote-targets`
        );
        return data.targets ?? [];
    },

    async getTarget(agentId: number, targetId: number): Promise<RemoteTarget> {
        return apiClient<RemoteTarget>(
            `${API}/${agentId}/remote-targets/${targetId}`
        );
    },

    async createTarget(
        agentId: number,
        payload: RemoteTargetCreate
    ): Promise<RemoteTarget> {
        return apiClient<RemoteTarget>(
            `${API}/${agentId}/remote-targets`,
            { method: "POST", json: payload }
        );
    },

    async updateTarget(
        agentId: number,
        targetId: number,
        payload: RemoteTargetUpdate
    ): Promise<RemoteTarget> {
        return apiClient<RemoteTarget>(
            `${API}/${agentId}/remote-targets/${targetId}`,
            { method: "PUT", json: payload }
        );
    },

    async deleteTarget(agentId: number, targetId: number): Promise<void> {
        await apiClient(`${API}/${agentId}/remote-targets/${targetId}`, {
            method: "DELETE",
        });
    },

    async testTarget(
        agentId: number,
        targetId: number
    ): Promise<{ connected: boolean; error?: string; message?: string }> {
        return apiClient(
            `${API}/${agentId}/remote-targets/${targetId}/test`,
            { method: "POST" }
        );
    },

    async getRemoteInventory(agentId: number): Promise<RemoteInventory> {
        return apiClient<RemoteInventory>(
            `${API}/${agentId}/remote-inventory`
        );
    },
};

File: backend/app/schemas/agent_remote_target.py
"""
Mission Control Agent Remote Target Schemas
"""

from datetime import datetime

from pydantic import BaseModel, Field


class RemoteTargetCreate(BaseModel):
    name: str
    hostname: str
    protocol: str = "psremoting"
    port: int | None = None
    username: str
    password: str | None = None
    ssh_key: str | None = None
    enabled: bool = True
    tags: str | None = None
    notes: str | None = None
    target_plugins: str | None = None
    db_type: str = Field("postgresql", pattern="^(postgresql|mssql)$")
    column_case: str = Field("pascal", pattern="^(pascal|snake)$")


class RemoteTargetUpdate(BaseModel):
    name: str | None = None
    hostname: str | None = None
    protocol: str | None = None
    port: int | None = None
    username: str | None = None
    password: str | None = None
    ssh_key: str | None = None
    enabled: bool | None = None
    tags: str | None = None
    notes: str | None = None
    target_plugins: str | None = None
    db_type: str | None = Field(None, pattern="^(postgresql|mssql)$")
    column_case: str | None = Field(None, pattern="^(pascal|snake)$")


class RemoteTargetResponse(BaseModel):
    id: int
    agent_id: int
    name: str
    hostname: str
    protocol: str
    port: int
    username: str
    enabled: bool
    tags: str | None = None
    notes: str | None = None
    last_collected_at: datetime | None = None
    last_status: str
    last_error: str | None = None
    created_at: datetime
    updated_at: datetime

    target_plugins: str | None = None
    db_type: str
    column_case: str

    model_config = {"from_attributes": True}


class AgentRemoteTargetsResponse(BaseModel):
    agent_id: int
    agent_name: str
    targets: list[RemoteTargetResponse]
